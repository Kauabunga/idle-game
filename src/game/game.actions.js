export const getClickAction = update => clicker =>
  update(state => {
    const { canClick } = clicker;
    // Check if can be clicked
    if (!canClick) {
      return state;
    }

    // Set clicked frame
    const clickerState = state.clickers[state.clickersMap[clicker.id]];
    clickerState.lastClickedFrame = state.tFrame;

    return state;
  });

export const getBuyAction = update => clicker =>
  update(state => {
    const { clickers, resources, tFrame, clickersMap } = state;
    const { canBuy, cost } = clicker;
    // Check if can buy
    if (!canBuy) {
      return state;
    }

    const clickerState = clickers[clickersMap[clicker.id]];
    clickerState.lastBuyFrame = tFrame;

    // Reduce resources
    Object.keys(cost).forEach(resourceCost => {
      resources[resourceCost] = (resources[resourceCost] || 0) - cost[resourceCost];
    });

    // increase count
    const buyCount = 1 * 1;
    resources[clicker.id] = (resources[clicker.id] || 0) + buyCount;

    return state;
  });
