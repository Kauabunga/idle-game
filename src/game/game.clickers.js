const baseClicker = ({ lastClickedFrame, createdAtFrame }) => ({
  count: 0,
  autoCount: 0,
  manualCount: 0,

  clickProgress: 0,

  lastClickedFrame: lastClickedFrame || null,
  createdAtFrame,
  createdAt: Date.now()
});

export const area01Level01 = options => ({
  ...baseClicker(options),
  id: "area01Level01",
  name: "First one",
  loadTime: 2000,
  cost: {
    workers: 10
  }
});

export const area01Level02 = options => ({
  ...baseClicker(options),
  id: "area01Level02",
  name: "Second one",
  loadTime: 3000,
  cost: {
    workers: 10,
    area01Level01: 10
  }
});

export const area01Level03 = options => ({
  ...baseClicker(options),
  id: "area01Level03",
  name: "Third one",
  loadTime: 5000,
  cost: {
    workers: 10,
    area01Level02: 20
  }
});

export const area01Level04 = options => ({
  ...baseClicker(options),
  id: "area01Level04",
  name: "Fourth one",
  loadTime: 7500,
  cost: {
    workers: 10,
    area01Level02: 50
  }
});
