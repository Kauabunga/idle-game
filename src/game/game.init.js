export const initState = {
  tFrame: 0,
  config: {
    workerLoadTime: 1000
  },

  multiplier: {
    workers: 1,

    area01: 1,
    area01Level01: 1,
    area01Level02: 1,
    area01Level03: 1,
    area01Level04: 1
  },

  resources: {
    workers: 0,

    area01: 0,
    area01Level01: 1,
    area01Level02: 1,
    area01Level03: 1,
    area01Level04: 1,

    area02: 0,
    area03: 0,
    area04: 0,
    area05: 0
  },
  clickers: []
};
