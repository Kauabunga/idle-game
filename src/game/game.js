import { writable } from "svelte/store";

import Loop from "./game.loop";
import gameUpdate from "./game.update";

function createGameStore() {
  const initState = { tFrame: 0, clickers: [] };
  const { subscribe, set, update } = writable({ tFrame: 0, clickers: [] });

  const handleClick = clicker =>
    update(state => {
      const { canClick } = clicker;
      // Check if can be clicked
      if (!canClick) {
        return state;
      }

      console.log(state);

      // Set clicked frame
      const clickerState = state.clickers[state.clickersMap[clicker.id]];
      clickerState.lastClickedFrame = state.tFrame;

      return state;
    });

  const handleBuy = clicker =>
    update(state => {
      const { canBuy } = clicker;
      // Check if can buy
      if (!canBuy) {
        return state;
      }

      // Set clicked frame
      const clickerState = state.clickers[state.clickersMap[clicker.id]];
      clickerState.lastBuyFrame = state.tFrame;

      return state;
    });

  return {
    subscribe,
    set,
    update,
    handleBuy,
    handleClick,
    reset: () => set(initState)
  };
}

export function Game() {
  const gameStore = createGameStore();

  const looper = new Loop(gameStore, gameUpdate);

  const addClicker = clickerFn =>
    gameStore.update(state => {
      const newClicker = clickerFn({ createdAtFrame: state.tFrame });
      return {
        ...state,
        clickers: [...state.clickers, newClicker],
        clickersMap: {
          ...state.clickersMap,
          [newClicker.id]: state.clickers.length
        }
      };
    });

  return {
    looper,
    addClicker,
    gameStore
  };
}
