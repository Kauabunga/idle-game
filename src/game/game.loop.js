/** Game Loop Module
 * This module contains the game loop, which handles
 * updating the game state and re-rendering the canvas
 * (using the updated state) at the configured FPS.
 */

export default function gameLoop(gameStore, update) {
  const loop = this;
  // Initialize timer variables so we can calculate FPS
  let before = window.performance.now();

  const fps = 60;
  const fpsInterval = 1000 / fps;
  // Set up an object to contain our alternating FPS calculations
  const cycles = {
    new: {
      frameCount: 0,
      startTime: before,
      sinceStart: 0
    },
    old: {
      frameCount: 0,
      startTime: before,
      sineStart: 0
    }
  };
  // Alternating Frame Rate vars
  const resetInterval = 5;
  let resetState = "new";

  loop.fps = 0;

  // Start off main loop
  loop.main = mainLoop;
  loop.main();

  return loop;

  // Main game rendering loop
  function mainLoop(tFrame) {
    // Request a new Animation Frame
    // setting to `stopLoop` so animation can be stopped via
    // `window.cancelAnimationFrame( loop.stopLoop )`
    loop.stopLoop = window.requestAnimationFrame(loop.main);

    // How long ago since last loop?
    const now = tFrame;
    const elapsed = now - before;
    let activeCycle;
    let targetResetInterval;

    // If it's been at least our desired interval, render
    if (elapsed > fpsInterval) {
      // Set before = now for next frame, also adjust for
      // specified fpsInterval not being a multiple of rAF's interval (16.7ms)
      // ( http://stackoverflow.com/a/19772220 )
      before = now - (elapsed % fpsInterval);

      // Increment the vals for both the active and the alternate FPS calculations
      for (var calc in cycles) {
        ++cycles[calc].frameCount;
        cycles[calc].sinceStart = now - cycles[calc].startTime;
      }

      // Choose the correct FPS calculation, then update the exposed fps value
      activeCycle = cycles[resetState];
      loop.fps = Math.round((1000 / (activeCycle.sinceStart / activeCycle.frameCount)) * 100) / 100;

      // If our frame counts are equal....
      targetResetInterval =
        cycles.new.frameCount === cycles.old.frameCount
          ? resetInterval * fps // Wait our interval
          : resetInterval * 2 * fps; // Wait double our interval

      // If the active calculation goes over our specified interval,
      // reset it to 0 and flag our alternate calculation to be active
      // for the next series of animations.
      if (activeCycle.frameCount > targetResetInterval) {
        cycles[resetState].frameCount = 0;
        cycles[resetState].startTime = now;
        cycles[resetState].sinceStart = 0;

        resetState = resetState === "new" ? "old" : "new";
      }

      // Update the game state
      // scope.state = scope.update(now);

      update(gameStore, tFrame);

      // Render the next frame
      // scope.render();
    }
  }
}
