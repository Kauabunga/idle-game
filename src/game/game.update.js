/** Game Update Module
 * Called by the game loop, this module will
 * perform any state calculations / updates
 * to properly render the next frame.
 */
export default function update(store, tFrame) {
  return store.update(state => {
    const { clickers } = state;

    const updatedClickers = clickers.map(clicker => updateClicker(clicker, tFrame));

    return {
      ...state,
      clickers: updatedClickers,
      tFrame
    };
  });
}

export function updateClicker(clicker, tFrame) {
  const { loadTime, createdAtFrame, canClick, lastClickedFrame, manualCount } = clicker;

  const nextAutoCount = (tFrame - createdAtFrame) / loadTime;

  // Compute can click state for displaying disabled state
  const clickFrameProgress = (lastClickedFrame && (tFrame - lastClickedFrame) / loadTime) || 0;

  const nextCanClickTime = clickFrameProgress > 1;
  const nextClickProgress = (!nextCanClickTime && clickFrameProgress % 1) || 0;
  let nextManualCount = manualCount || 0;

  let nextLastClickedFrame = lastClickedFrame;
  if (lastClickedFrame && !canClick && nextCanClickTime) {
    nextManualCount += 1;
    nextLastClickedFrame = null;
  }

  // const canClick = canClickTime && !auto
  const nextCanClick = !lastClickedFrame || nextCanClickTime;

  // Compute can buy state
  const nextCanBuy = true;

  const count = nextManualCount;
  return {
    ...clicker,
    count,
    autoCount: nextAutoCount,
    manualCount: nextManualCount,

    lastClickedFrame: nextLastClickedFrame,
    canClick: nextCanClick,
    clickProgress: nextClickProgress,

    canBuy: nextCanBuy
  };
}
