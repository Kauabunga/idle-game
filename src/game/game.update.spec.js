import { updateClicker } from './game.update';
import { area01Level01 } from './game.clickers';

// const lastClickedFrame = 1000;
// const updateClickerBefore = clicker => updateClicker(clicker, tFrame, config, resources, updateResources);

it('BASELINE', () => {
  const clicker = area01Level01({ createdAtFrame: 0 });
  const nextClicker = updateClicker(clicker, 0);

  expect(nextClicker.canClick).toBeTruthy();
  expect(nextClicker).toMatchSnapshot({ createdAt: expect.any(Number) });
});

describe('MANUAL', () => {
  describe('BEFORE click load', () => {
    it('Should not allow a click to be clicked if it has just been clicked', () => {
      const clicker = area01Level01({ createdAtFrame: 0, lastClickedFrame: 1000 });
      const nextClicker = updateClicker(clicker, clicker.loadTime + 999);
      expect(nextClicker.canClick).toBeFalsy();
    });
    it('Should allow another click once the load time passes', () => {
      const clicker = area01Level01({ createdAtFrame: 0, lastClickedFrame: 1000 });
      const nextClicker = updateClicker(clicker, clicker.loadTime + 1001);
      expect(nextClicker.canClick).toBeTruthy();
    });
  });

  describe('DURING click load', () => {
    it('Should update the progress after it has been clicked but not finished', () => {
      const clicker = area01Level01({ createdAtFrame: 0, lastClickedFrame: 1000 });
      const nextClicker = updateClicker(clicker, clicker.loadTime + 999);
      expect(nextClicker.clickProgress).toBeTruthy();
    });
  });

  describe('AFTER click load', () => {
    it('Should update the manual count after the click load has finished', () => {
      const clicker = area01Level01({ createdAtFrame: 0, lastClickedFrame: 1000 });
      const nextClicker = updateClicker(clicker, clicker.loadTime + 1001);
      expect(nextClicker.manualCount).toBeGreaterThan(clicker.manualCount);
    });
    it('Should reset progress after the click load has finished', () => {
      const clicker = area01Level01({ createdAtFrame: 0, lastClickedFrame: 1000 });
      const nextClicker = updateClicker(clicker, clicker.loadTime + 1001);
      expect(nextClicker.clickProgress).toBeFalsy();
    });
  });
});

// TODO: Auto
it('Should not allow a click to be clicked if it is in auto', () => {});
it('Should not allow a click to transition from manual to auto', () => {});
it('Should not allow a click to transition from different auto levels', () => {});
