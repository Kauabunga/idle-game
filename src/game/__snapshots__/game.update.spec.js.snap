// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BASELINE 1`] = `
Object {
  "autoCount": 0,
  "canBuy": true,
  "canClick": true,
  "clickProgress": 0,
  "cost": Object {
    "workers": 10,
  },
  "count": 0,
  "createdAt": Any<Number>,
  "createdAtFrame": 0,
  "id": "area01Level01",
  "lastClickedFrame": null,
  "loadTime": 2000,
  "manualCount": 0,
  "name": "First one",
}
`;
